app/forms/instance-create.tsx

```diff
@@ -160,29 +166,88 @@ export function CreateInstanceForm() {
+    projectImageSource: projectImages?.[0]?.id || '',
+    diskSource: disks?.[0]?.value || '',
     sshPublicKeys: allKeys,
-    // Use 2x the image size as the default boot disk size
-    bootDiskSize: Math.ceil(defaultImage?.size / GiB) * 2 || 10,
+    bootDiskSize: nearest10(defaultImage?.size / GiB),
   }
 
   const form = useForm({ defaultValues })
   const { control, setValue } = form
 
-  const imageInput = useWatch({ control: control, name: 'bootDiskSource' })
-  const image = allImages.find((i) => i.id === imageInput)
+  const bootDiskSourceType = useWatch({ control: control, name: 'bootDiskSourceType' })
+  const siloImageSource = useWatch({ control: control, name: 'siloImageSource' })
+  const projectImageSource = useWatch({ control: control, name: 'projectImageSource' })
+  const diskSource = useWatch({ control: control, name: 'diskSource' })
+  const bootDiskSource =
+    bootDiskSourceType === 'siloImage'
+      ? siloImageSource
+      : bootDiskSourceType === 'projectImage'
+        ? projectImageSource
+        : diskSource
+  const bootDiskSize = useWatch({ control: control, name: 'bootDiskSize' })
+  const image = allImages.find((i) => i.id === bootDiskSource)
   const imageSize = image?.size ? Math.ceil(image.size / GiB) : undefined
 
-  const sourceType = useWatch({ control: control, name: 'bootDiskSourceType' })
-
   useEffect(() => {
     if (createInstance.error) {
       setIsSubmitting(false)
     }
   }, [createInstance.error])
 
+  // additional form elements for projectImage and siloImage tabs
+  const bootDiskSizeAndName = (
+    <>
+      <div key="divider" className="!my-12 content-['a']" />
+      <DiskSizeField
+        key="diskSizeField"
+        label="Disk size"
+        name="bootDiskSize"
+        control={control}
+        min={imageSize || 1}
+        validate={(diskSizeGiB: number) => {
+          if (imageSize && diskSizeGiB < imageSize) {
+            return `Must be as large as selected image (min. ${imageSize} GiB)`
+          }
+        }}
+        disabled={isSubmitting}
+      />
+      <NameField
+        key="bootDiskName"
+        name="bootDiskName"
+        label="Disk name"
+        tooltipText="Will be autogenerated if name not provided"
+        required={false}
+        control={control}
+        disabled={isSubmitting}
+      />
+    </>
+  )
+
+  const createBootDisk = (values: InstanceCreateInput) => {
+    if (values.bootDiskSourceType === 'disk') {
+      return { type: 'attach' as const, name: values.diskSource }
+    }
+    const source =
+      values.bootDiskSourceType === 'siloImage'
+        ? values.siloImageSource
+        : values.projectImageSource
+    return {
+      type: 'create' as const,
+      name: values.bootDiskName || genName(values.name, source),
+      description: `Created as a boot disk for ${values.name}`,
+      size: values.bootDiskSize * GiB,
+      diskSource: { type: 'image' as const, imageId: source },
+    }
+  }
```

Comment:
Since as far as I can tell, this takes all relevant values as an argument, I would move it out of the component body. I would also rename it to `getBootDisk` or `getBootDiskAttachment` because `create` is easy to confuse with the `type: 'create' thing here. You can also put a return type of `InstanceDiskAttachment` on the function, more for clarity than anything else since it's checked regardless when you put it in the request body. It will also let you get rid of those `as const` bits, and gives slightly easier to understand type errors if we mess something up in here.